### ISSUE TITLE
control.getAttribute returns array of attributes for Form Component control

### ISSUE THREAD

**Issue by beyro** [MicrosoftDocs/powerapps-docs#3753]
Steps to reproduce:
* Add the Form Component to a lookup control on a form
* Call `formContext.getControl('lookupControlName').getAttribute()` on the lookup control.

Expected outcome:
* Per docs at: https://github.com/MicrosoftDocs/powerapps-docs/blob/main/powerapps-docs/developer/model-driven-apps/clientapi/reference/controls/getAttribute.md
    * getAttribute should return an attribute object.

Actual Outcome: 
* getAttribute returns an array of Attributes from the form that is referenced in the FormComponent control.

**Comment by JimDaly** [comment]
@beyro
Thanks for your feedback.

The feedback mechanism works best when you start from the published topic, rather than the source code in GitHub. This way the feedback gets routed to the owner based on the topic metadata and we can respond faster.

This topic: [control.getAttribute (Client API reference)](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/controls/getattribute) describes the method to find the attribute that is bound to the control. As you mention, this should be a single attribute.

I'm not sure what you mean by 'Form Component' or how you would apply it to a lookup.
Does it make sense that there should be multiple attributes returned?

If you want to report a bug, I recommend you contact technical support.
If you think there is a missing use case for this not covered in the docs, please explain more about what that is.

Thanks,

**Comment by beyro** [comment]
@JimDaly Thanks for the reply and your patience with my roundabout way of getting this feedback to you... I see now I should use the Feedback button at the bottom of the topic page - will use that next time!

This could be either a bug or a missed use case in the docs so I guess I'm looking for clarification of expected behavior.

When I say "Form Component Control", I mean the Form Component PCF control that is available with Dynamics 365 OOTB as described in the docs here:
https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/form-component-control

When you add this control to a lookup field, then calling control.getAttribute on it, returns an array of all the attributes on the <QuickForms> XML tag in the "Forms" property of the PCF control.

Based on the getAttribute docs, this is not expected behavior, but I thought maybe it is expected and the docs just need to be updated because it kind of makes sense to be able to access the attributes of the referenced form in this way.

**Comment by JimDaly** [comment]
@beyro 
Thanks for the explanation. I wasn't aware of the form component control.
That definitely doesn't seem like expected behavior.

The docs say: [Edit related table records directly from another table’s main form > Client API](https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/form-component-control#client-api)

> **Client API**
> A [form context](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/clientapi-form-context) is available for the form component control. It can be accessed via the main form's form context by accessing the control via an API, such as [getControl](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/controls/getcontrol). Before you access data for the related table in the form component control, event handlers should wait for the [isLoaded API](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/formcontext-ui-quickforms/isloaded) for the control to return true.

If I'm understanding you correctly, your code running on the main form and you are accessing the control using something like the sample here: [isLoaded (Client API reference) > Example](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/formcontext-ui-quickforms/isloaded#example)

```
function getAttributeValue(executionContext) {
    var formContext = executionContext.getFormContext();
    var quickViewControl = formContext.ui.quickForms.get("<QuickViewControlName>");
    if (quickViewControl != undefined) {
        if (quickViewControl.isLoaded()) {
            // Access the value of the column bound to the constituent control
            var myValue = quickViewControl.getControl(0).getAttribute().getValue();
            console.log(myValue);
            
            // Search by a specific column present in the control       
            var myValue2 =  quickViewControl.getControl().find(control => control.getName() == "<AttributeSchemaName>").getAttribute().getValue();
            console.log(myValue2);
            
            return;
        }
        else {
            // Wait for some time and check again
            setTimeout(getAttributeValue, 10, executionContext);
        }
    }
    else {
        console.log("No data to display in the quick view control.");
        return;
    }
}
```
The **Remarks** for **IsLoaded** says:

> The data binding for the constituent controls in a quick view control may not be complete during the main form OnLoad event because the quick view form that the control is bound to may not have loaded completely. As a result, using the [getAttribute](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/controls/getattribute) or any data-related methods on a constituent control might not work. The isLoaded method for the quick view control helps determine the data binding status for constituent controls in a quick view control.

If you haven't waited for the **IsLoaded** event, perhaps the multiple attributes is how it '.. might not work'?

Can you confirm you are waiting for **IsLoaded**?

**Comment by beyro** [comment]
@JimDaly 
Yes, I am waiting for `isLoaded` but I am not getting the control via `formContext.ui.quickForms.get` like in the example.
I am getting it while iterating through the controls of a section something like the example below. This script is running on a Case form and the Form Component control is linked to the Work Order Service Task main form:
```JavaScript
const section = formContext.ut.tabs.get('tabName').sections.get('sectionName')
section.controls.forEach(c => {
    if (c.isLoaded) {
        const attribute = c.getAttribute()
        // attribute is now a collection of attributes
        // The below outputs the attributes found on the Work Order Service Task Main form
        attribute.forEach(a => console.log(a.getName())) 
        // msdyn_surveyboundedoutput
        // msdyn_workorder
        // msdyn_customerasset
        // msdyn_workorderincident
        // msdyn_agreementbookingservicetask
        // msdyn_booking
        // msdyn_name
        // ownerid
        // msdyn_lineorder
    }
})
```
Interestingly, while the Form Component control does implement isLoaded like a QuickView control, it seems that Form Component controls are not part of the collection `formContext.ui.quickForms`.
Also, if you call `formComponentControl.getControlType`, it returns `'formcomponent'`, which is not one of the documented control types found here: [getControlType](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/controls/getcontroltype)

**Comment by JimDaly** [comment]
@beyro 
I've assigned this to @adrianorth because he is the author of [control.getAttribute (Client API reference)](https://learn.microsoft.com/en-us/power-apps/developer/model-driven-apps/clientapi/reference/controls/getattribute)

**Comment by JimDaly** [comment]
@beyro 

Have had a thread with the dev team and this is what we have so far:

1. A control can be associated with more than one attribute, so it seems it is possible for this method to return multiple attribute references. The control would have multiple attributes in the manifest. So we should update this topic to reflect that.

2. The Form Component control you are using is bound to a single lookup attribute, so it **should not** return multiple attributes. So, unless there is some other reason to expect multiple attributes to be bound to the control, you should contact technical support to report this.

**Comment by beyro** [comment]
@JimDaly @adrianorth 
Sorry for the delay in getting back to you on this.

It seems that getting the Form Component control via `formContext.getControl` is returning the formContext of the main form  referenced by the Form Component control.

If you call `formContext.getAttribute()` with no arguments from any main form, it also returns an array of all attributes on the form similar to how `formContext.,getControl('nameOfControlBoundToFormComponent').getAttribute() returns a collection of attributes.

In the [docs for the Form Component control](https://learn.microsoft.com/en-us/power-apps/maker/model-driven-apps/form-component-control#client-api), it says 
- A form context is available for the form component control. It can be accessed via the main form's form context by accessing the control via an API, such as getControl. Before you access data for the related table in the form component control, event handlers should wait for the isLoaded API for the control to return true.

I think this is saying that if you call `formContext.,getControl('nameOfControlBoundToFormComponent')`, it should return you the formContext of the bound main form. 

This seems to be true because calling
```JavaScript
formContext.getControl('nameOfControlBoundToFormComponent').getAttribute("msdyn_workorder").getValue()
```
works to get a specific attribute from the main form linked through the Form Component.

So I think this is behaving as expected, but the Form Component docs are not 100% clear - it would be great to have an example added to illustrate this part:
- "A form context is available for the form component control. It can be accessed via the main form's form context by accessing the control via an API, such as getControl"

**Comment by JimDaly** [comment]
Waiting on comment from @chmoncay.

**Comment by JimDaly** [comment]
Changes have been made to the following to address issues raised here:

https://learn.microsoft.com/power-apps/developer/model-driven-apps/clientapi/reference/controls/getcontroltype
https://learn.microsoft.com/power-apps/developer/model-driven-apps/clientapi/reference/controls
https://learn.microsoft.com/power-apps/developer/model-driven-apps/clientapi/reference/attributes

Thank you @chmoncay  and @fafuxa-ms !

Closing this.

**Comment by beyro** [comment]
Thanks everyone. The new docs make the form component control behavior much
clearer!
